src/util/syscall_sandbox.cpp

```diff
@@ -0,0 +1,488 @@
+    assert(SYSCALL_NAMES.size() == 333 && SYSCALL_NAMES[0] == "read" && SYSCALL_NAMES[332] == "statx" && "Syscalls must be listed in syscall number order without gaps.");
+    return syscall_number < SYSCALL_NAMES.size() ? SYSCALL_NAMES[syscall_number] : "*unknown*";
+}
+
+void SyscallSandboxDebugSignalHandler(int, siginfo_t* info, void* void_context)
+{
+    if (info->si_code != SYS_SECCOMP) {
+        return;
+    }
+    const ucontext_t* ctx = static_cast<ucontext_t*>(void_context);
+    if (ctx == nullptr) {
+        return;
+    }
+    std::set_new_handler(std::terminate);
+    const unsigned int syscall_number = ctx->uc_mcontext.gregs[REG_SYSCALL];
+    const std::string syscall_name = GetLinuxSyscallName(syscall_number);
+    const std::string thread_name = !util::ThreadGetInternalName().empty() ? util::ThreadGetInternalName() : "*unnamed*";
+    const std::string error_message = strprintf("ERROR: The syscall \"%s\" (syscall number %d) is not allowed by the syscall sandbox in thread \"%s\". Please report. Exiting.", syscall_name, syscall_number, thread_name);
+    tfm::format(std::cerr, "%s\n", error_message);
+    LogPrintf("%s\n", error_message);
+    std::terminate();
+}
+
+void InstallSyscallSandboxDebugHandler()
+{
+    static std::atomic<bool> syscall_reporter_installed{false};
```

Comment:
In commit "Add syscall sandboxing (seccomp-bpf)" (3636a5c1f03fd01da2a78cad0cba7ab69cadf074)

Use of std::atomic here does not seem to make sense. If multiple threads can call InstallSyscallSandboxDebugHandler for the first time simultaneously, then a `Mutex` and `LOCK` need to be used to prevent the later threads from returning before the first thread finishes calling `sigaction` and `sigprocmask`. If multiple threads can't call InstallSyscallSandboxDebugHandler for the first time simultaneously, then this can just be a normal bool instead of an atomic bool.

Reply:
Addressed: in the new code there is no need for locking at all since the setup is done in `SetupSyscallSandbox` which is guaranteed to only be called once. And you're absolutely right that `std::atomic<bool>` did not make much sense here :)